http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/util/ClusterHelper.scala

```diff
@@ -98,25 +113,204 @@ class ClusterHelper(dbRef: DbReference,
+          case Right((errors, op)) =>
+            Metrics.newRelic.incrementCounterIO("zoneCapacityClusterCreationFailure", errors.length)
+              .as(op)
+          case Left(errors) =>
+            Metrics.newRelic.incrementCounterIO("zoneCapacityClusterCreationFailure", errors.filter(whenGoogleZoneCapacityIssue).length)
+              .flatMap(_ => IO.raiseError(errors.head))
+        }
+
+        finalCluster = Cluster.createFinal(cluster, operation, stagingBucketName)
+
+      } yield (finalCluster, initBucketName, serviceAccountKeyOpt)
+
+      ioResult.handleErrorWith { throwable =>
+        cleanUpGoogleResourcesOnError(cluster.googleProject,
+          cluster.clusterName, initBucketName,
+          cluster.serviceAccountInfo, serviceAccountKeyOpt) >> IO.raiseError(throwable)
+      }
+    }
+  }
+
+  private def cleanUpGoogleResourcesOnError(googleProject: GoogleProject, clusterName: ClusterName, initBucketName: GcsBucketName, serviceAccountInfo: ServiceAccountInfo, serviceAccountKeyOpt: Option[ServiceAccountKey]): IO[Unit] = {
+    logger.error(s"Cluster creation failed in Google for $googleProject / ${clusterName.value}. Cleaning up resources in Google...")
+
+    // Clean up resources in Google
+    val deleteBucket = bucketHelper.deleteInitBucket(initBucketName) map { _ =>
+      logger.info(s"Successfully deleted init bucket ${initBucketName.value} for ${googleProject.value} / ${clusterName.value}")
+    } recover { case e =>
+      logger.error(s"Failed to delete init bucket ${initBucketName.value} for ${googleProject.value} / ${clusterName.value}", e)
+    }
+
+    // Don't delete the staging bucket so the user can see error logs.
+
+    val deleteCluster = IO.fromFuture(IO(gdDAO.deleteCluster(googleProject, clusterName))) map { _ =>
+      logger.info(s"Successfully deleted cluster ${googleProject.value} / ${clusterName.value}")
+    } recover { case e =>
+      logger.error(s"Failed to delete cluster ${googleProject.value} / ${clusterName.value}", e)
+    }
+
+    val deleteServiceAccountKey = removeServiceAccountKey(googleProject, serviceAccountInfo.notebookServiceAccount, serviceAccountKeyOpt.map(_.id)) map { _ =>
+      logger.info(s"Successfully deleted service account key for ${serviceAccountInfo.notebookServiceAccount}")
+    } recover { case e =>
+      logger.error(s"Failed to delete service account key for ${serviceAccountInfo.notebookServiceAccount}", e)
+    }
+
+    val removeIamRoles = removeClusterIamRoles(googleProject, serviceAccountInfo).map { _ =>
+      logger.info(s"Successfully removed IAM roles for ${googleProject.value} / ${clusterName.value}")
+    } recover { case e =>
+      logger.error(s"Failed to remove IAM roles for ${googleProject.value} / ${clusterName.value}", e)
+    }
+
+    List(deleteBucket, deleteCluster, deleteServiceAccountKey, removeIamRoles)
+      .map(Stream.eval)
+      .reduce(_ ++ _)
+      //.parJoin(4)  TODO: why does this not compile?
```

Comment:
I think you need Stream of Streams for `parJoin` to work. Here you have `List[Stream[IO, ???]`

Reply:
Yeah that makes sense... think I misunderstood `parJoin`